### DISCUSSION TITLE
Passport Strategy redirect_uri uses HTTP instead of HTTPS

### DISCUSSION THREAD

**Discussion by rafaelavd** [panva/openid-client#714] [Q&A] [answered]
Hi.

After upgrading to `openid-client` version `6.1.1` I am trying to authenticate against an Azure App Registration, following example https://github.com/panva/openid-client/blob/main/examples/passport.ts, with the only difference being on the callback endpoint:

```
app.get('/', (req, res, next) => {
  passport.authenticate(server.hostname, async function (err, passportUser, info) {
    res.end()
  })(req, res, next)
})
```

I get the following error (when invoking `passport.authenticate` above) to get the user from the data that Azure sends to the callback endpoint:

`AADSTS500112: The reply address 'http://[APP_URL]' does not match the reply address 'https://[APP_URL]' provided when requesting Authorization code. [...]`

notice the `http` on reply address instead of `https`.

The application is running on a cloud server on HTTPS. I made sure that App Registration on Azure has the `Redirect URI` defined to `https://[APP_URL]`.

If I switch to run the application locally on HTTP (http://localhost), and define `Redirect URI` to `http://localhost`, it works.

This is only happening since upgrading to `6.1.1`.

Since this is likely not to be a bug on `openid-client` I am posting as a question.

Thanks in advance for the help.

**Comment by panva** [comment]
Do you have your express app set to trust proxy and is your tls offloading proxy forwarding an x-forwarded-proto header?

**Reply by panva** [reply]
To explain, the `currentUrl` prototype method on the Strategy is used to determine the current URL the strategy is executed on, ergo the URL that is used as the `redirect_uri` token endpoint parameter. And in tls-offloading proxy setups it just depends on [express configuration](https://expressjs.com/en/guide/behind-proxies.html) to be in order.

https://github.com/panva/openid-client/blob/7405570e440ff39834d63d6a027a29ecd7dc537d/src/passport.ts#L378-L389

**Reply by panva** [reply, marked as answer]
So you have two options to remedy

- fix your proxy setup as per the express guide above (i.e. setting X-Forwarded-For, X-Forwarded-Host, and X-Forwarded-Proto headers at your reverse proxy)
- overwrite the `currentUrl` method on the Strategy instance to return the current URL

**Reply by rafaelavd** [reply]
Went with option 2 and it worked! Thanks.
However reading on the docs of `currentUrl` method above:

`[...] unless callbackURL was specified in the Strategy constructor`

I was in fact specifying `callbackURL` on the Strategy constructor to the correct URL `https://[APP_URL]`. So I am not sure there is an inconsistency on the documentation?

Thanks again for the quick support!

**Reply by panva** [reply]
Absolutely, the docs don't reflect the state of implementation atm. I'll have a look at it later.
